libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/BannerInstructionEvent.kt

```diff
@@ -1,34 +1,58 @@
 package com.mapbox.navigation.core.trip.session
 
 import com.mapbox.api.directions.v5.models.BannerInstructions
+import com.mapbox.navigation.utils.internal.ifNonNull
 
-internal class BannerInstructionEvent {
+internal class BannerInstructionEvent private constructor() {
 
-    var latestInstructionIndex: Int? = null
+    var latestInstructionWrapper: LatestInstructionWrapper? = null
         private set
 
+    val latestInstructionIndex: Int?
+        get() = latestInstructionWrapper?.latestInstructionIndex
+
+    val latestBannerInstructions: BannerInstructions?
+        get() = latestInstructionWrapper?.latestBannerInstructions
+
     var bannerInstructions: BannerInstructions? = null
         private set
 
-    var latestBannerInstructions: BannerInstructions? = null
-        private set
+    companion object {
+        operator fun invoke(): BannerInstructionEvent = BannerInstructionEvent()
+    }
 
     fun isOccurring(bannerInstructions: BannerInstructions?, instructionIndex: Int?): Boolean {
         return updateCurrentBanner(bannerInstructions, instructionIndex)
     }
 
-    fun invalidateLatestBannerInstructions() {
-        latestBannerInstructions = null
-        latestInstructionIndex = null
+    fun invalidateLatestBannerInstructions(latestInstructionWrapper: LatestInstructionWrapper?) {
+        if (latestInstructionWrapper == this.latestInstructionWrapper) {
+            this.latestInstructionWrapper = null
+        }
     }
 
     private fun updateCurrentBanner(banner: BannerInstructions?, instructionIndex: Int?): Boolean {
         bannerInstructions = banner
         if (bannerInstructions != null && bannerInstructions!! != latestBannerInstructions) {
-            latestBannerInstructions = bannerInstructions
-            latestInstructionIndex = instructionIndex
+            latestInstructionWrapper =
+                LatestInstructionWrapper.createOrNull(instructionIndex, bannerInstructions)
             return true
         }
         return false
     }
+
+    data class LatestInstructionWrapper(
+        val latestInstructionIndex: Int,
+        val latestBannerInstructions: BannerInstructions
+    ) {
+        companion object {
+            fun createOrNull(
```

Comment:
discussed: keep it as technical debt, the code is not really changed anything, and only keep the same approach. Ideally, a wrapper must be used whenever it might be to provide all or nothing.